dask/array/random.py

```diff
@@ -360,124 +675,398 @@ def permutation(self, x):
+            dummy_p = state._backend.array([1]) if p is not None else p
+            meta = state._backend.random.RandomState().choice(1, size=(), p=dummy_p)
+        else:
+            raise ValueError("Unknown generator class")
+        len_a = a
+        if a < 0:
+            raise ValueError("a must be greater than 0")
+    else:
+        a = asarray(a)
+        a = a.rechunk(a.shape)
+        meta = a._meta
+        if a.ndim != 1:
+            raise ValueError("a must be one dimensional")
+        len_a = len(a)
+        dependencies.append(a)
+        a = a.__dask_keys__()[0]
+
+    # Normalize and validate `p`
+    if p is not None:
+        if not isinstance(p, Array):
+            # If p is not a dask array, first check the sum is close
+            # to 1 before converting.
+            p = np.asarray(p)
+            if not np.isclose(p.sum(), 1, rtol=1e-7, atol=0):
+                raise ValueError("probabilities do not sum to 1")
+            p = asarray(p)
+        else:
+            p = p.rechunk(p.shape)
+
+        if p.ndim != 1:
+            raise ValueError("p must be one dimensional")
+        if len(p) != len_a:
+            raise ValueError("a and p must have the same size")
+
+        dependencies.append(p)
+        p = p.__dask_keys__()[0]
+
+    if size is None:
+        size = ()
+    elif not isinstance(size, (tuple, list)):
+        size = (size,)
+
+    if axis != 0:
+        raise ValueError("axis must be 0 since a is one dimensinal")
+
+    chunks = normalize_chunks(chunks, size, dtype=np.float64)
+    if not replace and len(chunks[0]) > 1:
+        err_msg = (
+            "replace=False is not currently supported for "
+            "dask.array.choice with multi-chunk output "
+            "arrays"
+        )
+        raise NotImplementedError(err_msg)
+
+    return a, size, replace, p, axis, chunks, meta, dependencies
+
+
+def _wrap_func(
+    rng, funcname, *args, size=None, chunks="auto", extra_chunks=(), **kwargs
+):
+    """Wrap numpy random function to produce dask.array random function
+    extra_chunks should be a chunks tuple to append to the end of chunks
+    """
+    if size is not None and not isinstance(size, (tuple, list)):
+        size = (size,)
+
+    shapes = list(
+        {
+            ar.shape
+            for ar in chain(args, kwargs.values())
+            if isinstance(ar, (Array, np.ndarray))
+        }
+    )
+    if size is not None:
+        shapes.append(size)
+    # broadcast to the final size(shape)
+    size = broadcast_shapes(*shapes)
+    chunks = normalize_chunks(
+        chunks,
+        size,  # ideally would use dtype here
+        dtype=kwargs.get("dtype", np.float64),
+    )
+    slices = slices_from_chunks(chunks)
+
+    def _broadcast_any(ar, shape, chunks):
+        if isinstance(ar, Array):
+            return broadcast_to(ar, shape).rechunk(chunks)
+        if isinstance(ar, np.ndarray):
+            return np.ascontiguousarray(np.broadcast_to(ar, shape))
```

Comment:
Will there be no CuPy arrays being evaluated here and in conditionals below https://github.com/dask/dask/pull/9038/files#diff-bacbead89e1b64dbdf54cbe8878e9a52445945d2c2710b2a11edc65364c6b16dR925 and https://github.com/dask/dask/pull/9038/files#diff-bacbead89e1b64dbdf54cbe8878e9a52445945d2c2710b2a11edc65364c6b16dR940 ? Do they need to be special-cased?

Reply:
I don't think so although @rjzamora can give a more definite answer. In the meantime, added code to raise an error if we get an unexpected type 8a83a0f